apps/voting/contracts/Voting.sol

```diff
@@ -122,79 +130,74 @@ contract Voting is IForwarder, AragonApp {
         objectionPhaseTime = _objectionPhaseTime;
     }
 
-    /**
-     * TODO: Calculate gas spending and add hint for more efficient look up in the array if needed
-     */
     function setDelegate(address _delegate) public {
         require(_delegate != address(0), ERROR_ZERO_ADDRESS_PASSED);
 
         address msgSender = msg.sender;
         require(_delegate != msgSender, ERROR_SELF_DELEGATE);
 
-        address prevDelegate = delegates[msgSender];
+        address prevDelegate = delegates[msgSender].delegate;
         require(_delegate != prevDelegate, ERROR_DELEGATE_SAME_AS_PREV);
 
+        uint256 votingPower = token.balanceOfAt(msgSender, getBlockNumber64() - 1);
+        require(votingPower > 0, ERROR_NO_VOTING_POWER);
+
         if (prevDelegate != address(0)) {
             _removeDelegatedAddressFor(prevDelegate, msgSender);
         }
 
         _addDelegatedAddressFor(_delegate, msgSender);
-        delegates[msgSender] = _delegate;
+        delegates[msgSender] = Delegate(_delegate, uint96(delegatedVoters[_delegate].addresses.length - 1));
 
         emit DelegateSet(msgSender, prevDelegate, _delegate);
     }
 
-    /**
-     * TODO: Calculate gas spending and add hint for more efficient look up in the array if needed
-     */
     function removeDelegate() public {
         address msgSender = msg.sender;
-        address prevDelegate = delegates[msgSender];
+        address prevDelegate = delegates[msgSender].delegate;
         require(prevDelegate != address(0), ERROR_DELEGATE_NOT_SET);
 
         _removeDelegatedAddressFor(prevDelegate, msgSender);
-        delegates[msgSender] = address(0);
+        delegates[msgSender] = Delegate(address(0), 0);
 
         emit DelegateSet(msgSender, prevDelegate, address(0));
     }
 
     function _addDelegatedAddressFor(address _delegate, address _voter) internal {
-        require(_delegate != address(0), ERROR_ZERO_ADDRESS_PASSED);
-        require(_voter != address(0), ERROR_ZERO_ADDRESS_PASSED);
```

Comment:
why no zero checks? or are those used with `require`s before that every time?

Reply:
The `_delegate` is checked in require [before](https://github.com/katamarinaki/aragon-apps/pull/2/files/337b031e11ad56f5a228d1fe94e9012d195d69f4#diff-b6a08f95413f3092266f49a1084493af6d2596fc3118b0cede9af64dd8d3c402R134) call, the `_voter` is `msg.sender` in the only one call of this fn, so it's not worth checking imo.